File: clients_java/client/src/org/openamq/client/state/StateWaiter.java
package org.openamq.client.state;

import org.apache.log4j.Logger;
import org.openamq.AMQException;

/**
 * Waits for a particular state to be reached.
 *
 * @author Robert Greig (robert.j.greig@jpmorgan.com)
 */
public class StateWaiter implements StateListener
{
    private static final Logger _logger = Logger.getLogger(StateWaiter.class);

    private final AMQState _state;

    private volatile boolean _newStateAchieved;

    private volatile Throwable _throwable;

    private final Object _monitor = new Object();

    public StateWaiter(AMQState state)
    {
        _state = state;
    }

    public void waituntilStateHasChanged() throws AMQException
    {
        synchronized (_monitor)
        {
            //
            // The guard is required in case we are woken up by a spurious
            // notify().
            //
            while (!_newStateAchieved && _throwable == null)
            {
                try
                {
                    _logger.debug("State not achieved so waiting...");
                    _monitor.wait();
                }
                catch (InterruptedException e)
                {
                    ;
                }
            }
        }

        if (_throwable != null)
        {
            _logger.debug("Throwable reached state waiter: " + _throwable);
            if (_throwable instanceof AMQException)
            {
                throw (AMQException) _throwable;
            }
            else
            {
                throw new AMQException("Error: "  + _throwable, _throwable);
            }
        }
    }

    public void stateChanged(AMQState oldState, AMQState newState)
    {
        synchronized (_monitor)
        {
            if (_logger.isDebugEnabled())
            {
                _logger.debug("stateChanged called");
            }
            if (_state == newState)
            {
                _newStateAchieved = true;

                if (_logger.isDebugEnabled())
                {
                    _logger.debug("New state reached so notifying monitor");
                }
                _monitor.notifyAll();
            }
        }
    }

    public void error(Throwable t)
    {
        synchronized (_monitor)
        {
            if (_logger.isDebugEnabled())
            {
                _logger.debug("exceptionThrown called");
            }

            _throwable = t;
            _monitor.notifyAll();
        }
    }
}

File: dotnet/xmsclient/Client/State/StateWaiter.cs
using System;
using System.Threading;
using log4net;

namespace OpenAMQ.XMS.Client.State
{
    public class StateWaiter : IStateListener
    {
         private static readonly ILog _logger = LogManager.GetLogger(typeof(StateWaiter));

        private readonly AMQState _state;

        private volatile bool _newStateAchieved;

        private volatile Exception _exception;

        private ManualResetEvent _resetEvent = new ManualResetEvent(false);
        
        public StateWaiter(AMQState state)
        {
            _state = state;
        }

        public void StateChanged(AMQState oldState, AMQState newState)
        {            
            if (_logger.IsDebugEnabled)
            {
                _logger.Debug("stateChanged called");
            }
            if (_state == newState)
            {
                _newStateAchieved = true;

                if (_logger.IsDebugEnabled)
                {
                    _logger.Debug("New state reached so notifying monitor");
                }
                _resetEvent.Set();
            }            
        }

        public void Error(Exception e)
        {            
            if (_logger.IsDebugEnabled)
            {
                _logger.Debug("exceptionThrown called");
            }

            _exception = e;
            _resetEvent.Set();
        }
        
        public void WaituntilStateHasChanged()
        {            
            //
            // The guard is required in case we are woken up by a spurious
            // notify().
            //
            while (!_newStateAchieved && _exception == null)
            {                    
                _logger.Debug("State not achieved so waiting...");
                _resetEvent.WaitOne();                    
            }

            if (_exception != null)
            {
                _logger.Debug("Throwable reached state waiter: " + _exception);
                if (_exception is AMQException)
                {
                    throw _exception;
                }
                else
                {
                    throw new AMQException("Error: "  + _exception, _exception);
                }
            }
        }
    }
}
